silo-to-silo/contracts/src/SiloToSilo.sol

```diff
@@ -0,0 +1,280 @@
+            bytes(
+                string.concat(
+                    '{"receiver_id": "',
+                    receiverId,
+                    '", "amount": "',
+                    Strings.toString(amount),
+                    '", "msg": "',
+                    message,
+                    '"}'
+                )
+            ),
+            ONE_YOCTO,
+            FT_TRANSFER_CALL_NEAR_GAS
+        );
+
+        PromiseCreateArgs memory callback = near.auroraCall(
+            address(this),
+            abi.encodeWithSelector(this.ftTransferCallCallback.selector, msg.sender, token, amount),
+            NO_DEPOSIT,
+            BASE_NEAR_GAS
+        );
+
+        callFtTransfer.then(callback).transact();
+    }
+
+    function ftTransferCallCallback(address sender, IEvmErc20 token, uint128 amount) external onlyRole(CALLBACK_ROLE) {
+        uint128 transferredAmount = 0;
+        if (AuroraSdk.promiseResult(0).status == PromiseResultStatus.Successful) {
+            transferredAmount = _stringToUint(AuroraSdk.promiseResult(0).output);
+        }
+
+        uint128 refund_amount = amount - transferredAmount;
+        if (refund_amount > 0) {
+            balance[token][sender] += refund_amount;
+        }
+    }
+
+    function withdraw(IEvmErc20 token) external {
+        require(near.wNEAR.balanceOf(address(this)) >= ONE_YOCTO, "Not enough wNEAR balance");
+
+        string storage tokenAccountId = registeredTokens[token].nearTokenAccountId;
+        require(bytes(tokenAccountId).length > 0, "The token is not registered!");
+
+        uint128 senderBalance = balance[token][msg.sender];
+        require(senderBalance > 0, "The signer token balance = 0");
+
+        PromiseCreateArgs memory callWithdraw = _callWithoutTransferWNear(
+            near,
+            tokenAccountId,
+            "ft_transfer_call",
+            bytes(
+                string.concat(
+                    '{"receiver_id": "',
+                    siloAccountId,
+                    '", "amount": "',
+                    Strings.toString(senderBalance),
+                    '", "msg": "',
+                    _addressToString(msg.sender),
+                    '"}'
+                )
+            ),
+            ONE_YOCTO,
+            WITHDRAW_NEAR_GAS
+        );
+
+        PromiseCreateArgs memory callback = near.auroraCall(
+            address(this),
+            abi.encodeWithSelector(this.withdrawCallback.selector, msg.sender, token),
+            NO_DEPOSIT,
+            BASE_NEAR_GAS
+        );
+
+        callWithdraw.then(callback).transact();
+    }
+
+    function withdrawCallback(address sender, IEvmErc20 token) external onlyRole(CALLBACK_ROLE) {
+        require(
+            AuroraSdk.promiseResult(0).status == PromiseResultStatus.Successful,
+            "ERROR: The `Withdraw` XCC is fail"
+        );
+
+        uint128 transferredAmount = _stringToUint(AuroraSdk.promiseResult(0).output);
+        balance[token][sender] -= transferredAmount;
```

Comment:
Let's decrease balance in `withdraw` and increase in `withdrawCallback` only if something goes wrong.